File: 2022/day03/python/solutionA.py
#!/bin/env python

input = open('../input.txt').read()

def format_input(input: str) -> 'list[str]':
    return input.split()

def solution(rucksacks:'list[str]') -> int:
    priorities = []
    
    for rucksack in rucksacks:
        first = set(rucksack[:len(rucksack) // 2])
        second = set(rucksack[len(rucksack) // 2:])
        item = first.intersection(second).pop()
        priorities.append(ord(item) - (ord('a') - 1 if item > 'Z' else 38))
    
    return sum(priorities)

print(solution(format_input(input)))

File: 2022/day03/ruby/solutionA.rb
#!/bin/env ruby
# frozen_string_literal: true

require 'set'

input = File.read('../input.txt')

# @param {String} input
# @return {Array(String)}
def format(input)
  input.split
end

# @param {Array(String)} rucksacks
# @return {Integer}
def solution(rucksacks)
  rucksacks.map! do |rucksack|
    first = rucksack[...rucksack.length / 2].chars.to_set
    second = rucksack[rucksack.length / 2..].chars.to_set
    item = first.intersection(second).first
    item.ord - (item > 'Z' ? 'a'.ord - 1 : 38)
  end.sum
end

print solution(format(input))
